### PR TITLE
Remove nodes slice in loop of takeByTopology

### PR DESCRIPTION
**Pull request by tedyu**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
In the first loop of ManagerImpl#takeByTopology, nodes slice is used to record node Ids.

This PR drops this slice and populates perNodeDevices map directly.

Reading the remainder of the code in ManagerImpl#takeByTopology, we can see that -1 is used to represent 'without topology'. I kept its handling the same with original code.

Since the traversal over Topology.Nodes (line 694) potentially covers all nodes in the cluster, the savings are two fold: in reducing one round of traversal and in saving unnecessary slice allocation.

**Which issue(s) this PR fixes**:
Fixes #

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Review by mattjmcnaughton — COMMENTED** [review]
@tedyu could you please share more about the benefits you see this change providing? Do we have evidence that this extra memory allocation for `var nodes` is causing actual issues for users?

If not, I think my preference would be to leave this code as it is. I personally find the new code a little less readable. Specifically, it appears that `perNodeDevices[-1]` is being overloaded and made to track state in a way it wasn't really designed for (i.e. `-1` is not a node).

Curious to hear from others as well :)

**Comment by tedyu** [comment]
Reading the remainder of the code in ManagerImpl#takeByTopology, we can see that -1 is used to represent 'without topology'. I kept its handling the same with original code.

Since the traversal over Topology.Nodes (line 694) potentially covers all nodes in the cluster, the savings are two fold: in reducing one round of traversal and in saving unnecessary slice allocation.

**Comment by mattjmcnaughton** [comment]
> Reading the remainder of the code in ManagerImpl#takeByTopology, we can see that -1 is used to represent 'without topology'. I kept its handling the same with original code.
> 
> Since the traversal over Topology.Nodes (line 694) potentially covers all nodes in the cluster, the savings are two fold: in reducing one round of traversal and in saving unnecessary slice allocation.

Thanks for the additional context! I suppose one additional option could be using a descriptive variable name instead of -1? I.e. `withoutTopology := -1` and then going from there? I think that would help address some of my concerns. What do you think?

**Comment by tedyu** [comment]
@mattjmcnaughton 
Thanks for the suggestion.

withoutTopology was taken - it is a slice in the same func.
I used nodeWithoutTopology to represent -1.

Please take another look.

File: pkg/kubelet/cm/devicemanager/manager.go [hunk 1]
```diff
@@ -688,23 +688,27 @@ func (m *ManagerImpl) takeByTopology(resource string, available sets.String, aff
...
-				perNodeDevices[node] = sets.NewString()
-			}
-			perNodeDevices[node].Insert(d)
+		if !nodeAdded {
+			perNodeDevices[nodeWithoutTopology].Insert(d)
 		}
 	}
+	if perNodeDevices[nodeWithoutTopology].Len() == 0 {
+		delete(perNodeDevices, -1)
+	}
```
Comment (klueska) [inline pkg/kubelet/cm/devicemanager/manager.go]: I agree with this change in general. However, for readability, I would prefer the following.
It makes the two cases symmetrical and easier to follow and verify its correctness.

```
        nodeWithoutTopology := -1
        for d := range available {
                if m.allDevices[resource][d].Topology == nil || len(m.allDevices[resource][d].Topology.Nodes) == 0 {
                        if _, ok := perNodeDevices[nodeWithoutTopology]; !ok {
                                perNodeDevices[nodeWithoutTopology] = sets.NewString()
                        }
                        perNodeDevices[nodeWithoutTopology].Insert(d)
                        continue
                }

                for _, node := range m.allDevices[resource][d].Topology.Nodes {
                        if _, ok := perNodeDevices[int(node.ID)]; !ok {
                                perNodeDevices[int(node.ID)] = sets.NewString()
                        }
                        perNodeDevices[int(node.ID)].Insert(d)
                }
        }
```

Comment (klueska) [inline pkg/kubelet/cm/devicemanager/manager.go]: I agree with this change in general. However, for readability, I would prefer the following.
It makes the two cases symmetrical and easier to follow and verify its correctness.

```
        nodeWithoutTopology := -1
        for d := range available {
                if m.allDevices[resource][d].Topology == nil || len(m.allDevices[resource][d].Topology.Nodes) == 0 {
                        if _, ok := perNodeDevices[nodeWithoutTopology]; !ok {
                                perNodeDevices[nodeWithoutTopology] = sets.NewString()
                        }
                        perNodeDevices[nodeWithoutTopology].Insert(d)
                        continue
                }

                for _, node := range m.allDevices[resource][d].Topology.Nodes {
                        if _, ok := perNodeDevices[int(node.ID)]; !ok {
                                perNodeDevices[int(node.ID)] = sets.NewString()
                        }
                        perNodeDevices[int(node.ID)].Insert(d)
                }
        }
```

Comment (tedyu) [reply pkg/kubelet/cm/devicemanager/manager.go]: @klueska 
Thanks for chiming in.

Your comment has been addressed.

**Review by mattjmcnaughton — COMMENTED** [review]
Going to let @klueska give the final confirmation, as he is worked more closely with this code than I have :)

Thanks for being open to feedback @tedyu :)

**Pull request merged**